src/tracing/metric_event.h

```diff
@@ -0,0 +1,181 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <lib/core/CHIPError.h>
+#include <tracing/metric_keys.h>
+#include <tracing/metric_macros.h>
+#include <tracing/registry.h>
+
+namespace chip {
+namespace Tracing {
+
+/**
+ * Define a metric that can be logged. A metric consists of a key-value pair. The value is
+ * currently limited to simple scalar values.
+ *
+ * Additionally a metric is tagged as either an instant event or marked with a begin and end
+ * for the event. When the latter is used, a duration can be associated between the two events.
+ */
+class MetricEvent
+{
+public:
+    MetricEvent(const MetricEvent &)             = default;
+    MetricEvent(MetricEvent &&)                  = default;
+    MetricEvent & operator=(const MetricEvent &) = default;
+    MetricEvent & operator=(MetricEvent &&)      = default;
+
+public:
+    // This specifies the different categories of metric events that can created. In addition to
+    // emitting an event, events paired with a kBeginEvent and kEndEvent can be used to track
+    // duration for the event. A kInstantEvent represents a one shot event.
+    enum class Type
+    {
+        kBeginEvent,  // Marks the begin of an event. This must be tied with an End event.
```

Comment:
why is begin tied to a value? what is a value for a begin?

Reply:
Overall I am not convinced this tie-in of metric with a begin/end/instant makes sense. I do not understand why some metric has to have a begin/end.

Reply:
To make it more specific: I believe we have to change this approach because `Start` overall has no metric value and for `end` we either have a metric (generally an error code) or no metric (e.g. for methods that just `return` without an error code).